File: src/main/java/org/jetbrains/bsp/bazel/server/bazel/utils/BazelArgumentsUtils.java
package org.jetbrains.bsp.bazel.server.bazel.utils;

import com.google.common.base.Joiner;
import java.util.List;
import java.util.stream.Collectors;
import org.jetbrains.bsp.bazel.server.bazel.params.BazelQueryKindParameters;

public final class BazelArgumentsUtils {

  private static final String JOIN_TARGETS_DELIMITER = " + ";
  private static final String FUNCTIONS_DELIMITER = " union ";

  private static final String MNEMONIC_COMMAND = "mnemonic";
  private static final String KIND_COMMAND = "kind";
  private static final String EXCEPT_COMMAND = "except";

  public static String getJoinedBazelTargets(List<String> targets) {
    String joinedTargets = joinBazelTargets(targets);

    return String.format("(%s)", joinedTargets);
  }

  public static String getMnemonicWithJoinedTargets(
      List<String> targets, List<String> languageIds) {
    String joinedTargets = joinBazelTargets(targets);

    return getMnemonicsForTargets(joinedTargets, languageIds);
  }

  private static String getMnemonicsForTargets(String targets, List<String> languageIds) {
    return languageIds.stream()
        .map(languageId -> getMnemonicForLanguageAndTargets(languageId, targets))
        .collect(Collectors.joining(FUNCTIONS_DELIMITER));
  }

  public static String getQueryKindForPatternsAndExpressions(
      List<BazelQueryKindParameters> parameters) {
    return parameters.stream()
        .map(BazelArgumentsUtils::getQueryKind)
        .collect(Collectors.joining(FUNCTIONS_DELIMITER));
  }

  public static String getQueryKindForPatternsAndExpressionsWithException(
      List<BazelQueryKindParameters> parameters, String exception) {
    String kind = getQueryKindForPatternsAndExpressions(parameters);
    return String.format("%s %s %s", kind, EXCEPT_COMMAND, exception);
  }

  public static String getQueryKind(BazelQueryKindParameters parameter) {
    return String.format("%s(%s, %s)", KIND_COMMAND, parameter.getPattern(), parameter.getInput());
  }

  private static String joinBazelTargets(List<String> targets) {
    return Joiner.on(JOIN_TARGETS_DELIMITER).join(targets);
  }

  private static String getMnemonicForLanguageAndTargets(String languageId, String targets) {
    return String.format("%s(%s, %s)", MNEMONIC_COMMAND, languageId, targets);
  }
}

File: bazelrunner/src/main/kotlin/org/jetbrains/bsp/bazel/bazelrunner/utils/BazelArgumentsUtils.kt
package org.jetbrains.bsp.bazel.bazelrunner.utils

import ch.epfl.scala.bsp4j.BuildTargetIdentifier
import org.jetbrains.bsp.bazel.bazelrunner.params.BazelQueryKindParameters

object BazelArgumentsUtils {
  private const val FUNCTIONS_DELIMITER = " union "
  private const val EXCEPT_COMMAND = "except"

  fun joinBazelTargets(
      includedTargets: List<BuildTargetIdentifier>,
      excludedTargets: List<BuildTargetIdentifier>
  ): String {
    val rawIncludedTargets = toRawUris(includedTargets)
    val rawExcludedTargets = toRawUris(excludedTargets)
    if (rawExcludedTargets.isEmpty()) {
      return getJoinedBazelTargets(rawIncludedTargets)
    }
    val joinedIncludedTargets = joinBazelTargets(rawIncludedTargets)
    val joinedExcludedTargets = rawExcludedTargets.joinToString(" - ")
    return "($joinedIncludedTargets - $joinedExcludedTargets)"
  }

  fun toRawUris(targets: List<BuildTargetIdentifier>): List<String> =
      targets.map { it.uri }

  fun getJoinedBazelTargets(targets: List<String>): String {
    val joinedTargets = joinBazelTargets(targets)
    return "($joinedTargets)"
  }

  fun calculateExcludedTargetsWithExcludedPrefix(targets: List<BuildTargetIdentifier>): List<String> =
      toRawUris(targets).map { "-$it" }

  fun getMnemonicWithJoinedTargets(
      targets: List<String>, languageIds: List<String>): String {
    val joinedTargets = joinBazelTargets(targets)
    return getMnemonicsForTargets(joinedTargets, languageIds)
  }

  private fun getMnemonicsForTargets(targets: String, languageIds: List<String>) =
      languageIds.joinToString(FUNCTIONS_DELIMITER) { getMnemonicForLanguageAndTargets(it, targets) }

  fun getQueryKindForPatternsAndExpressions(parameters: List<BazelQueryKindParameters>): String =
      parameters.joinToString(FUNCTIONS_DELIMITER) { getQueryKind(it) }

  fun getQueryKindForPatternsAndExpressionsWithException(
      parameters: List<BazelQueryKindParameters>, exception: String): String {
    val kind = getQueryKindForPatternsAndExpressions(parameters)
    return "$kind $EXCEPT_COMMAND $exception"
  }

  private fun getQueryKind(parameter: BazelQueryKindParameters) =
      "kind(${parameter.pattern}, ${parameter.input})"

  private fun joinBazelTargets(targets: List<String>) =
      targets.joinToString(" + ")

  private fun getMnemonicForLanguageAndTargets(languageId: String, targets: String) =
      "mnemonic($languageId, $targets)"
}
